### DISCUSSION TITLE
In Range Check for double / float values in generated c code

### DISCUSSION THREAD

**Discussion by mork2020** [cantools/cantools#573] [Ideas]
Can you give me feedback about the following idea? 
(If it make sense, i try to create a pull request.)

I want to use a range check on double/float values, before i pack them in a CAN message. E.g. I have the following element with a scale.
DBC-file:
`SG_ Acceleration : 4|16@1- (0.0002,0) [-6.5536|6.5534] "MeterPerSecondSquare"  Zone`

generated header file:
`/**`
`     * acceleration`
`     *`
`     * Range: -32768..32767 (-6.5536..6.5534 MeterPerSecondSquare)`
`     * Scale: 0.0002`
`     * Offset: 0`
`     */`
`    int16_t acceleration;`

The in rage method could look like this:
`bool xxxx_acceleration_is_in_phys_range(double value) {`
`    return ((value <= 6.5534) && (value >= -6.5536));`
`}`

One question is, how to handle rounding errors for double/floats in the range check?

**Comment by andlaus** [comment]
I suppose such methods could be convenient in some contexts. 

> One question is, how to handle rounding errors for double/floats in the range check?

you probably mean discretization errors? (i.e., what should happen if a given physical value lies inbetween two representable values?) I think these should simply be ignored because converting a physical into a raw value always comes with the implicit understanding of "up to the discretization accuracy": for example for a "Velocity" signal with a resolution of 0.1 km/h and an offset of -30 km/h, the discretized version of 123.456 km/h would be 1535 which corresponds to 123.5 km/h...) The main issue which I see here is what happens at the boundaries of the range because it is possible that the extrema cannot not be exactly represented using the discretized representation. (e.g. the maximum of above signal might be set to as 222.222 km/h, but there is no internal value which exactly represents that, only ones for 222.2 km/h and 222.3 km/h.) Note that this is also an issue in the python code...

**Reply by mork2020** [reply]
I added some lines to the the c_source.py file to generate the additional c code for the physical in range check. If this check is always generated, i have to adapt a lot of c code template test files (which is fine for me). Do you think we need a command line switch for this? If not my PR will include also a lot of changes in the c code template files for the tests.

One question/remark about the rounding, e.g. 32767 * 0.1 gets rounded by python to 3276.7000000000003 which seams fine (see https://www.h-schmidt.net/FloatConverter/IEEE754.html).

**Reply by andlaus** [reply]
> always generated, i have to adapt a lot of c code template test files (which is fine for me). Do you think we need a command line switch for this? If not my PR will include also a lot of changes in the c code template files for the tests.

you can do this by uncommenting [this line](https://github.com/cantools/cantools/blob/master/tests/test_command_line.py#L73) and run test_command_line. If this makes the diff large, so be it...

> One question/remark about the rounding, e.g. 32767 * 0.1 gets rounded by python to 3276.7000000000003 which seams fine (see https://www.h-schmidt.net/FloatConverter/IEEE754.html).

IMO, yes. I consider such small errors as "numerical noise" and all code should be robust to it...
